### PR TITLE
Adaptation of protein sequence dataset

### PR DESCRIPTION
**Pull request by jannisborn**
Concerning #32

### CODE REVIEW

**Comment by jannisborn** [comment]
Created a draft for a PR @annaweber209 so that it's easier to track the changes.
Looks great already to me! Some of the test methods are still uncommented, but great to see that we have the tests already.

@C-nit was mentioning it, and I agree it would be nice, to have the option to bring a pre-defined set of features to the language. Like taking pretrained embeddings, e.g. from a large-scale protein language model like @mfilipav is working on.

How should we coordinate this @C-nit @annaweber209? I could do it, but until ~Tuesday, I'm busy with other stuff.

File: pytoda/proteins/protein_feature_language.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        Args:
+            token_indexes (Indexes): a sequence of token indexes.
+
+        Returns:
+            str: an amino acid sequence representation.
+        """
+        return ''.join(
+            [
+                self.index_to_token.get(token_index, '')
+                for token_index in token_indexes
```
Comment (jannisborn) [inline pytoda/proteins/protein_feature_language.py]: This is might fail since  `token_index` is not `Indexes` but a list of Indexes.

**Comment by jannisborn** [comment]
PR ready for merge, maybe you can give it a quick look @drugilsberg @C-nit 

- Introduce ProteinFeatureLanguage, that inherits from ProteinLanguage and converts AA sequences into 2D Tensors.
- This is done according to 3 featurizations supported (arg names are: `float_features`, `binary_featues` and `blosum`).
- Introduced new Transforms `ListToTensor`, i.e. 2D extension of `ToTensor`: converts a list of tuples (where each tuple represents the featurization, in this case of an AA) into a 2D Tensor
- adapted ProteinSequenceDataset and ProteinProteinInteractionDataset to also deal with `ProteinFeatureLanguage` objects.
- added unittests for everything

**Review by C-nit — COMMENTED** [review]
Great work!

The one issue with the organisation I have is that anything `to_index` are bad names in the case of protein feature sequences that are not 1D.

I assume these are called by the Datasets, so they couldn't be changed?
If yes, this should lead us to consider separate Dataset classes. Do we want Datasets to conditionally return data of different dimensionality?

Now that I think about it, what is the benefit of having the dataset translate to features via the language vs including/doing this with a nn.embedding with set weights? More than the freedom to do it either way? Maybe we could include here also utility functions to create such nn.Embedding from the processing.py

Do you also think it could be desirable to combine 'binary' and 'features', or to provide your own dict of token to features?

The name ProteinFeatureLanguage is a bit ambiguous, protein features must not be on the amino acid level, but could also be 1D.

**Comment by jannisborn** [comment]
Thx for the review.
1. It's complicated to change the `token_to_index` for the reason you mention. I would refrain from tackling this now since it does not impair the behaviour of the package but rather is a style/code elegance question.
2. The benefit from the protein feature language are e.g. that you can transform back from the features directly to the sequence (`token_indexes_to_sequence`). You can also use the features anywhere outside of `nn.Embedding`, e.g. for non-NN-models. Moreover, `ProteinLanguage` and `ProteinFeatureLanguage` will serve as a starting point for byte-pair-encodings where you dynamically aggregate the most frequent substructures in order to reduce sequence length and decrease repetition of similar sequences. Finding this byte-pair encodings requires to keep track of occurrence statistics of the input files (`.smi` or `.fasta`), so it will be much easier to start from the PL/PFL.

**Comment by C-nit** [comment]
Usability outside of torch Datasets and `token_indexes_to_sequence` are good arguments for the feature language I was not considering. Explains the choice for tuples to contain the AA properties.

>rather is a style/code elegance question.

I disagree, `token_to_index` will be actively misleading for anybody new to the code. In python w/o types to help names matter even more.
In this case much is hardcoded anyway, so users of the code will not have to go look what goes on in the Language. Go ahead and merge. (But, you know, when I use a new library, I probably would.)

But these are just my thoughts on the matter. Output of actual indexes vs arbitrary valued vectors seems like sufficiently different Datasets. We should keep it in mind. When we need more conditionals and `isinstance` in the Dataset a refactor might be worth it then.

**Review by C-nit — APPROVED** [review]

**Pull request merged**